### DISCUSSION TITLE
Micronaut's UriMatchTemplate.match method doesn't work the same as how Spring's UriTemplate

### DISCUSSION THREAD

**Discussion by donbeave** [micronaut-projects/micronaut-core#4582] [Q&A] [answered]
I'm migrating my application from Spring Boot to Micronaut and found out that UriMatchTemplate is not exactly the same. Probably it's a bug, I am not sure yet, to figure it out I created this issue.

The main difference is that `match` method will return empty Optional instead of the result, here is an example of my code.

I have this URL format:

```java
private static final String URL_FORMAT = "https://www.mydomain.com/post/{postSlug}";
```

Spring Boot original code:

```java
UriTemplate template = new UriTemplate(MOVIE_URL_FORMAT);

Map<String, String> match = template.match("https://www.mydomain.com/post/abc");

// will return `abc`
match.get("postSlug");
```

I have converted this code to Micronaut:

```java
UriMatchTemplate template = UriMatchTemplate.of(URL_FORMAT);

// will return `Optional.empty`
Optional<UriMatchInfo> match = template.match("https://www.mydomain.com/post/abc")
```

I expect it will return `UriMatchInfo` object, where I can use `getVariableValues` method to retrieve the value for `postSlug` variable.

### Example Application

Please clone this repo: https://github.com/donbeave/micronaut-uritemplate, here is two tests which compare Micronaut and Spring behavior:
1) https://github.com/donbeave/micronaut-uritemplate/blob/master/app/src/test/java/micronaut/uritemplate/SpringTest.java
2) https://github.com/donbeave/micronaut-uritemplate/blob/master/app/src/test/java/micronaut/uritemplate/MicronautTest.java

**Comment by graemerocher** [comment, marked as answer]
It is not a design goal for `UriMatchTemplate` to be the same as the Spring impl. I believe you can use:

```
private static final String URL_FORMAT = "https://www.mydomain.com/post/{postSlug+}";
```

To achieve the same behaviour in Micronaut

**Reply by donbeave** [reply]
I thought both of these classes implement the same specification: https://tools.ietf.org/html/rfc6570#page-21

I tried `{postSlug+}` and got the same result:

<img width="1473" alt="image" src="https://user-images.githubusercontent.com/139017/100477443-4ff4ed00-3123-11eb-8f73-d35f19b0b5ef.png">

**Comment by graemerocher** [comment]
Apologies the + is actually before the path. See for examples https://github.com/micronaut-projects/micronaut-core/blob/74b67f092917e52b27b918f90cc05f30eb05ae78/http/src/test/groovy/io/micronaut/http/uri/UriMatchTemplateSpec.groovy#L72 

If anything IMO Micronaut implements the spec more accurately. See “Reserved Expansion” section https://tools.ietf.org/html/rfc6570#section-3.2.3

**Reply by donbeave** [reply]
Hmm, for some reason it doesn't work too:

<img width="1372" alt="image" src="https://user-images.githubusercontent.com/139017/100598259-73918080-3339-11eb-9a76-10e3ab24d17f.png">

**Reply by graemerocher** [reply]
Seems the current implementation only supports relative paths such as `/post/{+postSlug}` for matching (expansion works fine)

Feel free to open a feature request for absolute URIs

**Reply by donbeave** [reply]
Ok, thank you! I will do.
